### DISCUSSION TITLE
Using other libraries in circuits

### DISCUSSION THREAD

**Discussion by infinite-pursuits** [Consensys/gnark#671] [General]
Hi there,
I have some Golang code that I wish to create a GNARK circuit for. This code uses a priority queue and some custom data structs created by me. For priority queue, I am using a heap library that already exists in golang. I am also using the mat library for matrix and vector operations. Can I use these data structs and libraries in gnark as is?

**Comment by ivokub** [comment]
Unfortunately not. In-circuit all operations have to be performed using the given [circuit API](https://pkg.go.dev/github.com/consensys/gnark@v0.8.0/frontend#API).

So instead of
```go
func fn(a, b int) int {
    res := b+c
    res = res + 3
    return res
}
```
you have to define a circuit
```go
type Circuit struct {
    A, B frontend.Variable `gnark:",public"`
    Res frontend.Variable
}

func (c *Circuit) Define(api frontend.API) error {
    res := api.Add(c.A, c.B)
    res = api.Add(res, 3)
    api.AssertIsEqual(res, c.Res)
    return nil
}
```

You also have to keep in mind that if out-circuit int is usually 64 bits then in-circuit it is a large prime field element. So the operation results may be unexpected. To overcome this, we have implemented field emulation, but this comes at a performance cost.

Sometimes you can quite easily adapt out-circuit implementations into the circuit (for example, for Keccak hash function it was quite straightforward), but sometimes not. For example, for matrix multiplication it makes more sense to implement Freivalds algorithm in-circuit for efficiency.

**Comment by ivokub** [comment]
FYI - converted issue to discussion as is more suitable.

**Comment by infinite-pursuits** [comment]
Thanks! So I cannot use any libraries. All the functions have to be implemented using the circuit API.
1) Can I do something similar to function calls in the circuits? For eg. if I write code for matrix multiplication, do I have to insert it inline everytime or can I define another circuit that I can call within another circuit or something similar ?
2) Are there some gnark implementations of stack/queue/heap etc. that I could use?
3) Are there some forums for gnark where I can go to for debugging?

**Reply by ivokub** [reply]
1. You can implement functions. It is sufficient to pass `frontend.API` into the function a la [like here](https://pkg.go.dev/github.com/consensys/gnark@v0.8.0/std/permutation/keccakf#Permute).
2. There isn't, but I'm not really sure you actually would need it. Usually you use these data structures for efficient memory reads/writes, but arithmetic circuits work differently. When you write a circuit in gnark, then essentially you are writing a big mathematical equation which can be uniquely solved. So there really isn't a concept of a "memory", but rather how to compute a value from different inputs based on selector. Lets say you want to choose between value `v1` and `v2` based on bit value `b`. You compute it as `res = v1 * b + v2 * (1-b)`. See [logderivlooku](https://github.com/ConsenSys/gnark/blob/develop/std/lookup/logderivlookup/doc_test.go) and [selector](https://github.com/ConsenSys/gnark/tree/develop/std/selector) packages for pre-existing implementation to select between many values.
3. We don't have anything beyond Github discussions, sorry.

**Comment by infinite-pursuits** [comment]
Thanks! Can I call this permute function or some other function inside a circuit?

**Reply by ivokub** [reply]
Yep, you can. You would call it like this:
```go
func (c *Circuit) Define(api frontend.API) error {
    res := keccakf.Permute(api, c.Input)
    // do stuff with res
}
```

**Comment by infinite-pursuits** [comment]
Also I have the code in Golang. Is it possible to directly get a circuit from this code? Are there any libraries which can help me? Or Do I have to write the circuit using frontend myself?

**Reply by ivokub** [reply]
You define your circuit as a type which implements `frontend.Circuit` (i.e. it has method `Define(frontend.API) error`). You get the circuit after you run `frontend.Compile` on it. The compiled type is a private type to allow iterating on it, but it can be serialised/deserialized etc.
